docs/reference/transfers.md

```diff
@@ -185,10 +185,11 @@ Constraints:
 
 This is the interval in seconds after a
 [`pending`](#flagspending) transfer's [arrival at the cluster](#timestamp)
-that it may be posted or voided. Zero denotes absence of timeout.
+that it may be [posted](#flagspost_pending_transfer) or [voided](#flagsvoid_pending_transfer).
+Zero denotes absence of timeout.
 
 If the timeout expires and the pending transfer has not already been
-posted or voided, the pending transfer is voided automatically.
+posted or voided, the pending balance is removed automatically.
 
 Non-pending transfers cannot have a timeout.
 
```

Comment:
The scan order is by the `expires_at` index, sorted by `{ timestamp + timeout_ns, timestamp }`.

Reply:
It's also relevant to document that we can only expire 8190 transfers per `pulse`, so the user might still see pending balances for expired transfers if their request is pipelined together with a pulse in such condition.

Reply:
> so the user might still see pending balances for expired transfers if their request is pipelined together with a pulse in such condition.

Ahh I hadn't figured that out yet, but that makes perfect sense! The pulse is injected into the _end_ of the pipeline, so by the time it commits, there may be `pipeline_prepare_queue_max - 1` prepares queued after it, so if there are a lot of expired transfers then the pulse might not hit all of them. Or it does hit all of them, but some more happen to expire while those trailing prepares are being processed, before the next pulse can expire them.

Reply:
Also, because we can only determine if we need the next pulse after the current one has been prefetched, that's why we support only one pulse at a time in the pipeline.

We can improve this logic in the future, replacing `expire_pending_transfers_pulse_timestamp` with some kind of "histogram" mapping multiple timestamps with a resolution of seconds to determine if we need to inject multiple pulses in the same pipeline, solving this issue.